cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractRunRecordCorrectorService.java

```diff
@@ -83,17 +85,21 @@ private void validateAndCorrectRunningRunRecords(final ProgramType programType,
                                                    final Set<String> processedInvalidRunRecordIds) {
     final Map<RunId, ProgramRuntimeService.RuntimeInfo> runIdToRuntimeInfo = runtimeService.list(programType);
 
+    com.google.common.base.Predicate<RunRecordMeta> filter = new com.google.common.base.Predicate<RunRecordMeta>() {
+      @Override
+      public boolean apply(RunRecordMeta input) {
+        String runId = input.getPid();
+        // Check if it is not actually running.
+        return !runIdToRuntimeInfo.containsKey(RunIds.fromString(runId));
+      }
+    };
     LOG.trace("Start getting run records not actually running ...");
-    Collection<RunRecordMeta> notActuallyRunning =
-      store.getRuns(ProgramRunStatus.RUNNING,
-                    new com.google.common.base.Predicate<RunRecordMeta>() {
-                      @Override
-                      public boolean apply(RunRecordMeta input) {
-                        String runId = input.getPid();
-                        // Check if it is not actually running.
-                        return !runIdToRuntimeInfo.containsKey(RunIds.fromString(runId));
-                      }
-                    }).values();
+    List<RunRecordMeta> notActuallyRunning = new ArrayList<>();
+
+    notActuallyRunning.addAll(store.getRuns(ProgramRunStatus.RUNNING, filter).values());
```

Comment:
should we also check SUSPENDED?

Reply:
+1 It is a good idea to check all non-final states

Reply:
scenario to reproduce would be similar: do a `yarn application -kill <app-id>` on a suspended workflow.

Reply:
Related to that, it seems that the transition table does not allow going from RESUMING to FAILED. https://github.com/caskdata/cdap/blob/release/4.3/cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/AppMetadataStore.java#L107

Reply:
I'm going to open a separate jira for suspended correct to limit the changes this close to the release. Suspended is an existing bug and I would imagine its far less common. I didn't even know you could suspend a workflow until now.